### ISSUE TITLE
Interface contract breaks with existence of optional properties, and compiler is unable to spot it and type check passes

### ISSUE THREAD

**Issue by subodhsgod** [microsoft/TypeScript#51039]
# Bug Report
If interface has optional property (one or more), and consumer passes value for un-existed properties on interface, compiler is unable to detect it, especially if the value object is defined as “const”. However, compiler could detect compile time Typecheck error if value object is “inline”. This is very inconsistent behavior. In both cases compiler should detect Typecheck error. For details pl. see code section.

### 🔎 Search Terms
interface optional property contract Typecheck 

### 🕗 Version & Regression Information
- I spotted this on 3-Oct-2022, but it existed in our code since long.
- Also tried to reproduce it using Typescript Playground using v4.8.4 and Nightly - v4.9.0-dev.20221003.
- This is the behavior in every version I tried, and I reviewed the FAQ for entries about Typecheck, contract, optional property.

### ⏯ Playground Link
[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.8.4&ssl=46&ssc=7&pln=41&pc=1#code/JYOwLgpgTgZghgYwgAgHIFcC2AjaBnZAbwChlkAHKAe3IEEAuZELXKAGlIuvICFHmc0ANydKNAMIB+fi2HEAvsWIB6ZcgCyEMAAsqAE2TACVEABsAnsgRxycbKZRUYyOHr2gA5skxwAHsgBmJlkoPDZkMAB3KhcoFB8QPTgwKihzcICoAyNkGjBgEzhTYgQTPDAXNwxBUOQAXmQACgFWPEZq1oBKeoA+Ik4HCri8dFMKhpb8ADoxOmQAamCavBnuHk5gGGaQldnxbpIyMmHRivmJndWJETJFMlKQPCoHKdMqD0aAcgAlCBGxz7hE5jToiRQqNTiOB4FAARkY4iomHIwAcUGQ0SgAGsCNDkBBfOQIAhIHopiUyhUAG5FWH1fpkWYMZCwjiMtaMABMHEUrj0HXwjRpplhoKUqmQUJhyE5CKRKLRGNSOJcBAJRJJEDJFMe1KKnPphy4NGZrNEHJlbON5HEjACCmIfIFoSF+rFEMl0JQATlyNR0GQeiofyYVAqmJVePVxNJ4TgiUMBBAYZc9hQKWQeHIKYAKuYNdpiVj8VBqFByQ9yshhUEGkamYwzeyaHxLeaJHarbMACKMAAsPMdVR2rtMAXdEqlKD7jAAqiBSphMBBwMhcG9IlZ9OmYmA0oYwFNJfL-eiI7i1YSY1q4wmcnYHBEYlnc-niYWEMXoGXyRKnSOSAlZs6EbDggOtVtuQ9YDbUCMC1GA3tkAHD15FBIA)

### 💻 Code
```ts
interface Numbers {
  propA: number,
  propB: number;
  propC?: number;
}

// Method is only capable of adding max 3 numbers, two are mandatory, 3rd is optional
const addNumbers = (numbers: Numbers) => {
  let result = numbers.propA + numbers.propB
  if(numbers.propC) {
    result += numbers.propC;
  }
  console.log('Result', result);
}

// Case 1: Compiler works as expected.
const val1 = {
  propA: 1,
  propB: 2,
}
addNumbers(val1);

// Case 2: Compiler works as expected.
const val2 = {
  propA: 1,
  propB: 2,
  propC: 3
}
addNumbers(val2);

// Case 3: Compiler does not works as expected, and is not able to spot Typecheck error.
const val3 = {
  propA: 1,
  propB: 2,
  propC: 3,
  propD: 4,
}
addNumbers(val3);

// Case 4: Uncomment below code to try it. Compiler works as expected, and is able to spot Typecheck error.
// addNumbers({
//   propA: 1,
//   propB: 2,
//   propC: 3,
//   propD: 4,
// });
```

### 🙁 Actual behavior
Compiler is not detecting is as Typecheck.

### 🙂 Expected behavior
Compiler should detect Typecheck error especially, as Compiler is able to detect it as compile time error when object is passed as inline.

**Comment by fatcerberus** [comment]
There is no type error, and this has nothing to do with optional properties.  Types are not exact; objects can have extra properties not mentioned in their type.  TypeScript checks for excess properties when you directly pass an object literal (because then you’d immediately lose access to the extra props), but this is not a type check *per se*; it’s more of a lint check.  See #12936.

**Comment by subodhsgod** [comment]
Thanks @fatcerberus for explanation.
Hmm, so by default Typescript allows any interface to have extra properties than what is defined! So `Numbers` interface above behaves almost like below from Typescript perspective, and for Typescript there is no difference! :-(
`
interface Numbers {
  propA: number,
  propB: number;
  propC: number;
  [key: string]: number | string | object | Date // etc...
}
`

I really wish Type safety could be little stricter without depending upon linter! and validation with object literal could be extended to all objects, may be with optional config setting in `tsconfig`.

**Comment by subodhsgod** [comment]
With my last comment, the issue got closed

**Comment by subodhsgod** [comment]
Thanks @fatcerberus.
